File: platforms/android/module/src/main/kotlin/rs/whisker/runtime/ModuleDefinitionSamples.kt
// Compile-time smoke checks for the `ModuleDefinition` DSL.
//
// These samples exist purely so a break in the DSL surface fails
// `compileDebugKotlin` here. `internal` so they stay out of the
// public module ABI. No runtime behavior — they just exercise the
// type surface end-to-end.

package rs.whisker.runtime

@Suppress("unused", "TestFunctionName")
internal object ModuleDefinitionSamples {

    // ---- View-bearing module --------------------------------------------

    internal class FakeVideoView {
        fun setSrc(value: String) { /* noop */ }
        fun play() { /* noop */ }
        fun pause() { /* noop */ }
        fun seek(seconds: Double) { /* noop */ }
        fun setFontSize(value: Float) { /* noop */ }
    }

    internal fun videoModuleDefinition(): ModuleDefinition = ModuleDefinition {
        Name("Video")

        View("sample-module:Video", FakeVideoView::class.java) {
            Prop("src") { view: FakeVideoView, value ->
                view.setSrc(value.asString() ?: "")
            }
            Command("play") { view: FakeVideoView, _ -> view.play() }
            Command("pause") { view: FakeVideoView, _ -> view.pause() }
            Command("seek") { view: FakeVideoView, parameters ->
                view.seek(parameters.asDouble() ?: 0.0)
            }
            Events("onCompleted")
            TextStyle { view: FakeVideoView, style -> view.setFontSize(style.fontSize) }
            Measurement { request ->
                WhiskerMeasuredSize(request.knownWidth ?: 160f, request.knownHeight ?: 90f)
            }
        }
    }

    // ---- Function-only (view-less) module ------------------------------

    internal fun localStoreModuleDefinition(): ModuleDefinition = ModuleDefinition {
        Name("WhiskerLocalStore")

        Function("save") { args ->
            val key = args.getOrNull(0)?.asString() ?: ""
            val value = args.getOrNull(1)?.asString() ?: ""
            WhiskerValue.Bool(key.isNotEmpty() && value.isNotEmpty())
        }
        Function("load") { args ->
            WhiskerValue.Str("stub-value-for-${args.getOrNull(0)?.asString() ?: ""}")
        }
    }

    // ---- Module subclass shape -----------------------------------------

    @WhiskerModule
    internal class StubModule : Module() {
        override fun definition(): ModuleDefinition = videoModuleDefinition()
    }
}

File: platforms/ios/Sources/WhiskerModule/ModuleDefinitionSamples.swift
// Compile-time smoke checks for the `ModuleDefinition` DSL.
//
// These samples exist purely so a break in the DSL surface fails
// `swift build` here. No runtime behavior — they just exercise the
// type surface end-to-end.

import Foundation

internal enum ModuleDefinitionSamples {

    // MARK: - View-bearing module

    internal class FakeVideoView {
        func setSrc(_ value: String) { /* noop */ }
        func play() { /* noop */ }
        func pause() { /* noop */ }
        func seek(_ seconds: Double) { /* noop */ }
        func setFontSize(_ value: CGFloat) { /* noop */ }
    }

    internal static func videoModuleDefinition() -> ModuleDefinition {
        ModuleDefinition {
            Name("Video")

            View("sample/Video", FakeVideoView.self) {
                Prop("src") { (view: FakeVideoView, value: WhiskerValue) in
                    view.setSrc(value.asString ?? "")
                }
                Command("play")  { (view: FakeVideoView, _: WhiskerValue) in view.play() }
                Command("pause") { (view: FakeVideoView, _: WhiskerValue) in view.pause() }
                Command("seek")  { (view: FakeVideoView, parameters: WhiskerValue) in
                    view.seek(parameters.asDouble ?? 0)
                }
                Events("onCompleted")
                TextStyle { (view: FakeVideoView, style: WhiskerTextStyle) in
                    view.setFontSize(style.fontSize)
                }
                Measurement { request in
                    WhiskerMeasuredSize(
                        width: request.knownWidth ?? 160,
                        height: request.knownHeight ?? 90
                    )
                }
            }
        }
    }

    // MARK: - Function-only (view-less) module

    internal static func localStoreModuleDefinition() -> ModuleDefinition {
        ModuleDefinition {
            Name("WhiskerLocalStore")

            Function("save") { (args: [WhiskerValue]) -> WhiskerValue in
                let key = args.first?.asString ?? ""
                let value = args.count > 1 ? (args[1].asString ?? "") : ""
                return .bool(!key.isEmpty && !value.isEmpty)
            }
            Function("load") { (args: [WhiskerValue]) -> WhiskerValue in
                .string("stub-value-for-\(args.first?.asString ?? "")")
            }
        }
    }

    // MARK: - WhiskerModule subclass shape

    @WhiskerModule
    internal final class StubModule: Module {
        public override func definition() -> ModuleDefinition {
            ModuleDefinitionSamples.videoModuleDefinition()
        }
    }
}
